CI: Build Fail on raspi

## Maintainer Summary
2026-06-17 03:06:02 UTC

## Mea Cupla
- About 2x weeks ago
  - ordered_map 

## Host
```
Linux raspberrypi 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64 GNU/Linux
Debian clang version 11.0.1-2
Target: aarch64-unknown-linux-gnu
```

## Build
```
mkdir pr-1400
cd pr-1400
git clone https://github.com/InternationalColorConsortium/iccDEV.git
cd iccDEV/Build
git fetch origin pull/1400/head:pr-1400
git checkout pr-1400
git branch && echo "---" && git log --oneline -1
CC=clang CXX=clang++ CXXFLAGS="-fsanitize=address,undefined,integer,bounds,null,float-divide-by-zero,alignment,vla-bound -fno-omit-frame-pointer -g -O0" LDFLAGS="-fsanitize=address,undefined,integer,bounds,null,float-divide-by-zero,alignment,vla-bound" cmake Cmake -DCMAKE_BUILD_TYPE=Debug -DENABLE_ASAN=ON -DENABLE_UBSAN=ON -DENABLE_TOOLS=ON -DENABLE_WXWIDGETS=OFF
make -j32
        cd ../Testing/
        echo "=== Updating PATH ==="
         for d in ../Build/Tools/*; do
          [ -d "$d" ] && export PATH="$(realpath "$d"):$PATH"
         done
```

## Error
```
[ 81%] Built target iccApplyToLink
In file included from /home/xss/pr-1400/iccDEV/IccConnect/IccLibConnect/IccCmmConfig.cpp:68:
In file included from /home/xss/pr-1400/iccDEV/Build/Cmake/IccConnect/../../../IccConnect/IccLibConnect/IccCmmConfig.h:1:
In file included from /usr/bin/../lib/gcc/aarch64-linux-gnu/10/../../../../include/c++/10/list:60:
In file included from /usr/bin/../lib/gcc/aarch64-linux-gnu/10/../../../../include/c++/10/bits/stl_algobase.h:64:
/usr/bin/../lib/gcc/aarch64-linux-gnu/10/../../../../include/c++/10/bits/stl_pair.h:218:11: error: field has incomplete type 'nlohmann::basic_json<unordered_map, std::vector, std::__cxx11::basic_string<char>, bool, long, unsigned long, double, std::allocator, adl_serializer, std::vector<unsigned char, std::allocator<unsigned char>>>'
      _T2 second;                ///< The second member
          ^
/usr/bin/../lib/gcc/aarch64-linux-gnu/10/../../../../include/c++/10/ext/aligned_buffer.h:91:28: note: in instantiation of template class 'std::pair<const std::__cxx11::basic_string<char>, nlohmann::basic_json<unordered_map, std::vector, std::__cxx11::basic_string<char>, bool, long, unsigned long, double, std::allocator, adl_serializer, std::vector<unsigned char, std::allocator<unsigned char>>>>' requested here
    : std::aligned_storage<sizeof(_Tp), __alignof__(_Tp)>
```

--- a/Build/Cmake/CMakeLists.txt
+++ b/Build/Cmake/CMakeLists.txt
@@ -458,7 +458,31 @@ option(ENABLE_ICCJSON             "Build IccJSON library support"               
 option(ENABLE_IMAGE_TOOLS         "Build TIFF/PNG/JPEG image-dependent tools"           ON)
 option(ENABLE_CMM_TOOLS           "Build Windows CMM DLL tool"                          ON)
 option(ENABLE_IIS_TOOLS           "Build Windows IIS ISAPI tool"                        ON)
-option(ICC_JSON_ORDERED           "Use ordered_json for deterministic key order (slower)" OFF)
+include(CheckCXXSourceCompiles)
+set(_ICC_JSON_ORDERED_DEFAULT OFF)
+
+set(_ICC_CHECK_GLIBCXX_SOURCE "
+#if defined(__has_include)
+#  if __has_include(<bits/c++config.h>)
+#    include <bits/c++config.h>
+#  endif
+#endif
+#if defined(_GLIBCXX_RELEASE) && _GLIBCXX_RELEASE < 11
+#  error libstdc++ before 11 cannot instantiate unordered basic_json here
+#endif
+int main() { return 0; }
+")
+check_cxx_source_compiles("${_ICC_CHECK_GLIBCXX_SOURCE}" _ICC_JSON_UNORDERED_STDLIB_OK)
+if(NOT _ICC_JSON_UNORDERED_STDLIB_OK)
+  set(_ICC_JSON_ORDERED_DEFAULT ON)
+elseif(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 11)
+  set(_ICC_JSON_ORDERED_DEFAULT ON)
+endif()
+
+option(ICC_JSON_ORDERED           "Use ordered_json for deterministic key order (auto-ON on libstdc++<11)" ${_ICC_JSON_ORDERED_DEFAULT})
+if(ICC_JSON_ORDERED AND _ICC_JSON_ORDERED_DEFAULT)
+  message(STATUS "ICC_JSON_ORDERED auto-enabled: libstdc++<11 cannot build unordered basic_json")
+endif()
 option(ICC_TRACE_NAN_ENABLED      "Enable tracing NaN inputs in debug builds"           OFF)
 option(ICC_CLUT_DEBUG             "Enable CLUT debugging support"                       OFF)
 option(ICC_ENABLE_ASSERTS         "Enable ICC_ASSERT traps and debug assertions"        OFF)

